[wip] use yarn workspace to build packages

xcaptain opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 952


## What is the new behavior?
use workspace to build packages, decrease node_modules size and makes debug much easier


## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
I upgraded typescript version and `@types/node` version, which I think is breaking change

## Other information

xcaptain commented:
I suggest delete the `bundle/` directory for 3 reasons:

1. `bundle/` contains out dated code like `packages/core/errors/exceptions/microservices-package-not-found.exception.d.ts` has been removed but still in `bundle/core/errors/exceptions/microservices-package-not-found.exception.d.ts`
2. `bundle/` contains duplicated code as `packages` but in version control which makes editor find much more difficult.
3. Using yarn workspace, the generated lib is in their own directory, so no need to use `bundle/` any more.

BrunnerLivio commented:
@xcaptain 
I second this. Already asked about that with issue #658, but did not get much response. 

Another reason is;

- Build artifacts from the `bundle` folder makes the repository unnecessary larger and pollutes the git history.

There are not many pros in my opinion. The only benefit Nest would loose, is the ability to use Nest Git repository as direct dependency of your application (not a NPM package):

`"repository": "github:user/repo"`

 But in my opinion that is something people should not do anyway.

xcaptain commented:
I think I'm almost done with this issue, current status:

`@nestjs/common` 97 test case passed
`@nestjs/core`  244 passed, 2 failed
`@nestjs/microservices` 214 passed, 20 failed
`@nestjs/websocket` 32 passed

At first I just want to restructure the project and use yarn workspace to manage packages, but then I can't help to update some dependencies versions, so resulting a lot of changes. 

## how to test this change

1. test per package
    `cd packages/common; yarn test`
2. test the whole workspace
    `yarn test`

## test the samples

I tested `sample/01-cats-app` using `yarn link`

1. `cd packages/common; yarn link`
2. `cd sample/01-cats-app; yarn link '@nestjs/common'`
...

after linked the 5 packages, then run `yarn install` and then `yarn serve`

BrunnerLivio commented:
@xcaptain maybe you should wait with fixing everything, until @kamilmysliwiec approves #952, so you did not do work for nothing?

BrunnerLivio commented on package.json:
```diff
@@ -2,24 +2,13 @@
-    "prepublish:rc": "npm run build:lib && npm run copy-docs",
-    "prepublish:next": "npm run build:lib && npm run copy-docs",
-    "prepublish:beta": "npm run build:lib && npm run copy-docs",
-    "publish": "./node_modules/.bin/lerna publish --exact -m \"chore(@nestjs) publish %s release\"",
```
Why delete the publish scripts? If there is a valid reason, you should atlest document how to publish in a markdown file I suppose.

xcaptain replied:
Thanks, publish is easy, just run `yarn run lerna publish`, I will add the detail publish script later.

kamilmysliwiec commented:
I really like the idea. Good job @xcaptain, keep rocking 🙂

xcaptain commented:
@kamilmysliwiec I'm planning to make these changes into version 5.3.0, I added publish script and fixed more tests since the last commit. current test status:

`@nestjs/common` 101 test case passed
`@nestjs/core` 252 passed
`@nestjs/microservices` 214 passed, 20 failed
`@nestjs/websocket` 32 passed

The `@nestjs/microservices` package is hard to test, I don't have so much dependencies install on my machine, so I will just skip these.

xcaptain commented:
closed in favor of https://github.com/nestjs/nest/pull/1060
